### PR TITLE
DFPL-702 Global search Migration

### PR DESCRIPTION
**Pull request by so99y**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/DFPL-702

### Change description ###
1. Global Search - Create top-level fields
2. Grant the read access of top-level fields to all users.
3. Set supplementary data via API callback
4. Data migration on case data to enable global search

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: ccd-definition/RoleToAccessProfiles.json [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+    "RoleName": "idam:caseworker-publiclaw-systemupdate",
+    "ReadOnly": "N",
+    "AccessProfiles": "caseworker-publiclaw-systemupdate,GS_profile"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "RoleName": "[LASOLICITOR]",
+    "ReadOnly": "N",
+    "AccessProfiles": "[LASOLICITOR]"
```
[the remark this answers is not in the document]
Comment (so99y) [reply ccd-definition/RoleToAccessProfiles.json:105]: Added back roles from `CaseRoles.json`. 

`[CREATOR]` which exists in `CaseRole.java` was not defined in `CaseRoles.json`.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/MigrateCaseController.java [hunk 2]
```diff
@@ -61,6 +79,70 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
...
     }
 
+    @PostMapping("/submitted")
+    public void handleSubmitted(@RequestBody CallbackRequest callbackRequest) {
+        final CaseData caseData = getCaseData(callbackRequest);
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+
+        String migrationId = (String) caseDetails.getData().get(MIGRATION_ID_KEY);
+
+        if ("DFPL-702".equals(migrationId)) {
```
Comment (DanCatchpole) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/MigrateCaseController.java:89]: we need to remove this, the field won't be present on the case data during the submitted callback, we should always perform this supplementary data update I think (and we will make sure that when this is merged in NO ONE performs migrations until we remove it again the day after release)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/MigrateCaseController.java [hunk 3]
```diff
@@ -61,6 +79,70 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
...
+        if (caseData.getOrders() != null && StringUtils.isNotEmpty(caseData.getOrders().getCourt())) {
+            courtCode = caseData.getOrders().getCourt();
+        } else if (caseData.getCourt() != null) {
+            courtCode = caseData.getCourt().getCode();
+        }
+        if (courtCode == null) {
+            log.warn("Migration {id = DFPL-702, case reference = {}, case state = {}} doesn't have court info "
+                    + "therefore unable to set caseManagementLocation which is mandatory in global search.",
+                caseId, caseData.getState().getValue());
+            return;
```
Comment (DanCatchpole) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/MigrateCaseController.java:116]: think this should throw our normal Exception so it pops up immediately in the migration tool and is counted

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/MigrateCaseController.java [hunk 4]
```diff
@@ -61,6 +79,70 @@ public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody Cal
...
+
+            caseDetails.getData().put("caseNameHmctsInternal", caseName);
+            caseDetails.getData().put("caseManagementCategory", DynamicList.builder()
+                .value(DynamicListElement.builder().code("FPL").label("Family Public Law").build())
+                .listItems(List.of(
+                    DynamicListElement.builder().code("FPL").label("Family Public Law").build()
+                ))
+                .build());
+        } else {
+            log.warn("Migration {id = DFPL-702, case reference = {}, case state = {}} fail to lookup ePIMMS ID "
```
Comment (DanCatchpole) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/MigrateCaseController.java:135]: also throw that exception here to keep track

**Review by DanCatchpole — DISMISSED** [review]
lgtm!

**Pull request merged**